inductor: add input type check for fuse_attention

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99502
* __->__ #99296


For TIMM ```xcit_large_24_p8_224```, the scale factor is a tensor(https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/xcit.py#L205), and ```scaled_dot_product_attention``` doesn't support it, this PR will add a check which only does the fusion when the scale factor is float/int value.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor: add input type check for fuse_attention

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: add input type check for fuse_attention"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: add input type check for fuse_attention"



For TIMM ```xcit_large_24_p8_224```, the scale factor is a tensor(https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/xcit.py#L205), and ```scaled_dot_product_attention``` doesn't support it, this PR will add a check which only does the fusion when the scale factor is float/int value.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented on test/inductor/test_fused_attention.py:
```diff
@@ -165,6 +165,42 @@ def sfdp_pattern_6(query, key, value):
         self._check_common(sfdp_pattern_6, contains=False)
 
+    def test_pattern_fails_with_tensor_factor(self):
+        # https://github.com/pytorch/pytorch/issues/99124
```
This issue is reported on CPU side, but I checked it is also happening on GPU side, so add this case here.

EikanWang commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
+        scale_factor_node = list(view_node.users.keys())[0]
+        if len(scale_factor_node.args) != 2:
+            return False
+        # make sure the scale_factor a float/int. SymInt?
```
Should we add a dynamic shape case to verify `SymInt`?

XiaobingSuper replied:
Yes, we may need to add dynamic shape case, but the patterns doesn't match for dynamic shape case. Let do it at next step.

jansel commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
+        # make sure the mul(div) for the scale factor is scalar mul(div).
+        # bmm->view->mul(div)
+        matmuls = filter_nodes(match.nodes, aten.bmm)
+        if len(matmuls) < 2:
```
Shouldn't this already be checked by the pattern?  Can you give an example and add a test for when this fails?

XiaobingSuper replied:
Yes, this has been checked by the pattern, I do that to make sure it is all right. I will remove it to simplify it.

jansel commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
+        if len(matmuls) < 2:
+            return False
+        if (
+            len(matmuls[0].users) != 1
```
Shouldn't this already be checked by the pattern?  Can you give an example and add a test for when this fails?

XiaobingSuper replied:
removed such a check.

jansel commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
+            return False
+        view_node = list(matmuls[0].users.keys())[0]
+        if (
+            len(view_node.users) != 1
```
Shouldn't this already be checked by the pattern?  Can you give an example and add a test for when this fails?

XiaobingSuper replied:
removed such a check.

jansel commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
+            or list(view_node.users.keys())[0].target != scale_factor_op
+        ):
+            return False
+        scale_factor_node = list(view_node.users.keys())[0]
```
Could you go directly here with `filter_nodes(match.nodes, scale_factor_op)`?

XiaobingSuper replied:
Yes, changed.

jansel commented on test/inductor/test_fused_attention.py:
```diff
@@ -165,6 +165,42 @@ def sfdp_pattern_6(query, key, value):
+            torch.randn(tensor_shape, device="cuda"),
+            torch.randn(tensor_shape, device="cuda"),
+        ]
+        with torch.no_grad():
```
Can you test training as well?

XiaobingSuper replied:
training test is added, please note that even if not have this PR, the training path can work well(the pattern doesn't match), and has an accuracy gap compared with eager mode.

XiaobingSuper committed:
Update on "inductor: add input type check for fuse_attention"



For TIMM ```xcit_large_24_p8_224```, the scale factor is a tensor(https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/xcit.py#L205), and ```scaled_dot_product_attention``` doesn't support it, this PR will add a check which only does the fusion when the scale factor is float/int value.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: add input type check for fuse_attention"



For TIMM ```xcit_large_24_p8_224```, the scale factor is a tensor(https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/xcit.py#L205), and ```scaled_dot_product_attention``` doesn't support it, this PR will add a check which only does the fusion when the scale factor is float/int value.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: add input type check for fuse_attention"



For TIMM ```xcit_large_24_p8_224```, the scale factor is a tensor(https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/xcit.py#L205), and ```scaled_dot_product_attention``` doesn't support it, this PR will add a check which only does the fusion when the scale factor is float/int value.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
